src/mangrove/query_builder.hpp

```diff
@@ -1132,79 +1131,169 @@ class bit_update_expr {
 };
 
 /* Query comparison operators */
+// Note that each comparison operator has two versions:
+// a) One in which a field is compared to a value of the same type, and
+// b) One in which an array field is compared to a value of the contained type.
+//    (e.g. an array of int's can be compared to a single integer)
+//    This is syntactic sugar for MongoDB's $elemMatch operator, and mirrors
+//    the syntax in the actual query language.
 
 template <typename NvpT, typename = std::enable_if_t<is_nvp_v<NvpT>>>
 constexpr comparison_expr<NvpT, typename NvpT::no_opt_type> eq(
     const NvpT &lhs, const typename NvpT::no_opt_type &rhs) {
     return {lhs, rhs, "$eq"};
 }
 
+template <typename NvpT, typename = std::enable_if_t<is_nvp_v<NvpT>>,
+          typename = std::enable_if_t<is_iterable_v<typename NvpT::no_opt_type>>>
+constexpr comparison_expr<NvpT, typename NvpT::array_element_type> eq(
+    const NvpT &lhs, const typename NvpT::array_element_type &rhs) {
+    return {lhs, rhs, "$eq"};
+}
+
 template <typename NvpT, typename = std::enable_if_t<is_nvp_v<NvpT>>>
 constexpr comparison_expr<NvpT, typename NvpT::no_opt_type> operator==(
     const NvpT &lhs, const typename NvpT::no_opt_type &rhs) {
     return eq(lhs, rhs);
 }
 
+template <typename NvpT, typename = std::enable_if_t<is_nvp_v<NvpT>>,
+          typename = std::enable_if_t<is_iterable_v<typename NvpT::no_opt_type>>>
+constexpr comparison_expr<NvpT, typename NvpT::array_element_type> operator==(
+    const NvpT &lhs, const typename NvpT::array_element_type &rhs) {
+    return eq(lhs, rhs);
+}
+
 template <typename NvpT, typename = std::enable_if_t<is_nvp_v<NvpT>>>
 constexpr comparison_expr<NvpT, typename NvpT::no_opt_type> gt(
     const NvpT &lhs, const typename NvpT::no_opt_type &rhs) {
     return {lhs, rhs, "$gt"};
 }
 
+template <typename NvpT, typename = std::enable_if_t<is_nvp_v<NvpT>>,
+          typename = std::enable_if_t<is_iterable_v<typename NvpT::no_opt_type>>>
+constexpr comparison_expr<NvpT, typename NvpT::array_element_type> gt(
```

Comment:
It's awkward, because you actually want both no_opt_type and array_element_type to be allowed. Another level of confusion is that use std::is_same on T will cause issues, since passing string literals (const char[]) to std::string fields will fail is_same. std::is_convertible also is problematic because passing `int32` to an `int64` field will be pass is_convertible, but then end up inserting an int32 into what should be an int64, and cause de-serialization errors. 

I might be able to do something like:
```
template <typename NvpT, typename U,
          typename = std::enable_if_t<is_nvp_v<NvpT>>,
          typename R = std::conditional<is_convertible<U, array_element_type>, array_element_type,
                                        no_opt_type>>
constexpr comparison_expr<NvpT, R> gt(const NvpT& lhs, const U& rhs);
```

plus or minus some `_v`'s and `::value`'s here and there. 
The idea is that the first enable_if restricts the type of U loosely, using `is_convertible`, but then `R` is chosen to be the appropriate Nvp type, so that the comparison_expr knows what type it's dealing with. (and doesn't accidentally send int32's to int64's and etc. )